zebra-state/src/service/memory_state.rs

```diff
@@ -460,11 +402,199 @@ impl Ord for Chain {
+                self.chain_set.insert(chain);
+            }
+            // else discard `chain`
+        }
+
+        // return the finalized block
+        finalized_block
+    }
+
+    /// Commit block to the non-finalize state.
+    pub fn commit_block(&mut self, block: Arc<Block>) {
+        let parent_hash = block.header.previous_block_hash;
+
+        let mut parent_chain = self
+            .take_chain_if(|chain| chain.non_finalized_tip_hash() == parent_hash)
+            .or_else(|| {
+                self.chain_set
+                    .iter()
+                    .find_map(|chain| chain.fork(parent_hash))
+                    .map(Box::new)
+            })
+            .expect("commit_block is only called with blocks that are ready to be commited");
+
+        parent_chain.push(block);
+        self.chain_set.insert(parent_chain);
+    }
+
+    pub fn commit_new_chain(&mut self, block: Arc<Block>) {
+        let mut chain = Chain::default();
+        chain.push(block);
+        self.chain_set.insert(Box::new(chain));
+    }
+
+    pub fn best_chain_len(&self) -> block::Height {
+        block::Height(
+            self.chain_set
+                .iter()
+                .next()
+                .expect("only called after inserting a block")
+                .blocks
+                .len() as u32,
+        )
+    }
+
+    pub fn any_chain_contains(&self, hash: &block::Hash) -> bool {
+        self.chain_set
+            .iter()
+            .any(|chain| chain.height_by_hash.contains_key(hash))
+    }
+
+    /// Remove and return the first chain satisfying the given predicate.
+    fn take_chain_if<F>(&mut self, predicate: F) -> Option<Box<Chain>>
+    where
+        F: Fn(&Chain) -> bool,
+    {
+        let chains = mem::take(&mut self.chain_set);
+        let mut chains = chains.into_iter();
+
+        while let Some(chain) = chains.next() {
+            // if the predicate says we should remove it
+            if predicate(&chain) {
+                // add back the remaining chains
+                for chain in chains {
+                    self.chain_set.insert(chain);
+                }
+
+                // and return the chain
+                return Some(chain);
+            } else {
+                // add the chain back to the set and continue
+                self.chain_set.insert(chain);
+            }
+        }
+
+        None
+    }
+}
+
+/// A queue of blocks, awaiting the arrival of parent blocks.
+#[derive(Default)]
+pub struct QueuedBlocks {
+    /// Blocks awaiting their parent blocks for contextual verification.
+    blocks: HashMap<block::Hash, QueuedBlock>,
+    /// Hashes from `queued_blocks`, indexed by parent hash.
+    by_parent: HashMap<block::Hash, HashSet<block::Hash>>,
+    /// Hashes from `queued_blocks`, indexed by block height.
+    by_height: BTreeMap<block::Height, HashSet<block::Hash>>,
+}
+
+impl QueuedBlocks {
+    pub fn queue(&mut self, new: QueuedBlock) {
```

Comment:
> It might also help to explain that the CheckpointVerifier fulfils this condition, so it is able to use CommitFinalizedBlock. (But the BlockVerifier does not, so it must use CommitBlock.)

I'm not following this part. `QueuedBlocks` is only ever used for non-finalized blocks passed in via `CommitBlock`.

Reply:
Ah, right, I must have confused it with the similar code in sled_state.rs, which should require unique blocks. Function documentation might have helped here.

My overall concern here is that the function panics on duplicate blocks, but the BlockVerifier can't guarantee that it only approves each block once, because it doesn't have any state or context.

The current CommitBlock request design also doesn't handle all the edge cases here:
https://github.com/ZcashFoundation/zebra/blob/126488e7a3ce2bdfacad6556d99178351fc30423/book/src/dev/rfcs/0005-state-updates.md#committing-non-finalized-blocks

In particular, it needs to handle:
* duplicate non-finalised blocks
* duplicate queued blocks

Reply:
I think the latter check subsumes the former check but I agree, this is an extra edge case we need to handle. As far as I can tell this won't substantially change the design so I think it might be best to split this out into a separate issue that we fix in a follow up PR.

Reply:
All 3 checks are required, because a block can be in 3 different states: queued, non-finalized, and finalized.

So when a new block arrives, we need to check:
1. Is this block a duplicate of a block that is in the queue? 
2. Is this block a duplicate of a block that is in one of the non-finalized chains?
3. Is this block a duplicate of a block that is in the finalized chain?

Reply:
(I'll make a quick PR with these changes to the state RFC.)

Reply:
2 and 3 should be covered by the first step of `Committing non-finalized blocks` where it returns true if the block already exists, though it looks like we didn't specify that we should also check that the block isn't present in the non-finalized state.

Reply:
See https://github.com/ZcashFoundation/zebra/pull/1136 for the state RFC update.